### PR TITLE
Properly handle all Monitoring exporters being disabled

### LINKED ISSUE #40919: Explicitly disabling Monitoring exporters causes issues
Since 6.6, if you _explicitly_ disable all Monitoring exporters and enabled collection (both require explicit steps, although enabling collection can be done via the Monitoring UI), such as

```http
PUT /_cluster/settings
{
  "persistent": {
    "xpack.monitoring": {
      "collection.enabled": true,
      "exporters.my_exporter": {
        "type": "local",
        "enabled": false
      }
    }
  }
}
```

then this will trigger a bug where a key `ActionListener` is never given a response.

This will break Elasticsearch's monitoring service by never releasing its semaphore on the next collection; this does not trigger a memory leak, but it does stop any future monitoring collections until the node is restarted (without these conditions present). The symptom that gives this away is a total lack of monitoring data from Elasticsearch, but more directly you should find these messages in your `DEBUG`-level logs:

```
[2019-04-05T20:34:52,500][DEBUG][o.e.x.m.MonitoringService] [node-1] monitoring execution is skipped until previous execution terminated
```

In the same way, it will also never respond to `POST /_xpack/monitoring/_bulk` requests (any monitoring data from Kibana, Logstash, Beats, or APM Server), which causes their transport tasks to remain open (although the rest of the memory _is_ freed). The symptom that gives this away is that the tasks remain even though no monitoring threads are actually running:

```http
GET /_cat/tasks?v
action                              task_id                        parent_task_id                 type      start_time    timestamp running_time ip        node
cluster:admin/xpack/monitoring/bulk dh19vrTDTbWU9unnZD6dDA:2922459 -                              transport 1554508069093 23:47:49  47.7m        127.0.0.1 node-1
cluster:admin/xpack/monitoring/bulk dh19vrTDTbWU9unnZD6dDA:2922483 -                              transport 1554508070097 23:47:50  47.7m        127.0.0.1 node-1
```

### PR DESCRIPTION
**Pull request by pickypg**
When monitoring exporters are all disabled, which must be done explicitly, _and_ monitoring collection is enabled, then any call to `_xpack/monitoring/_bulk` will create a task that never closes _and_ ES collection will stop happening because a semaphore is never marked as completed.

This also simplifies the async `ExportBulk` code by removing the third step (second async step, `close`) entirely because it was entirely unnecessary by both implementations.

Closes #40919

### CODE REVIEW

**Comment by pickypg** [comment]
Ping @elastic/stack-monitoring

**Review by ycombinator — APPROVED** [review]
Functionally, this PR LGTM. I tested it as follows:

1. I enabled Monitoring collection and created a local exporter that was disabled:

   ```
   PUT _cluster/settings
   {
     "persistent": {
       "xpack.monitoring.exporters": {
         "my_remote": {
           "type": "local",
           "enabled": false
         }
       },
       "xpack.monitoring.collection.enabled": true
     }
   }
   ```

2. Then I started up Kibana so it would make requests to the `POST _monitoring/bulk` API every 10s.

3. I checked the list of running tasks with `GET _cat/tasks` and verified that there were no `cluster:admin/xpack/monitoring/bulk` tasks piling up without getting closed.

As for a code review, I'd really prefer someone from the @elastic/es-core-features team to take a look, maybe @jakelandis since he has been in the vicinity of ES monitoring code relatively recently?

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalBulk.java [hunk 1]
```diff
@@ -52,13 +52,10 @@
     }
 
     @Override
-    public void doAdd(Collection<MonitoringDoc> docs) throws ExportException {
+    protected void doAdd(Collection<MonitoringDoc> docs) throws ExportException {
         ExportException exception = null;
 
         for (MonitoringDoc doc : docs) {
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalBulk.java]: Do we still need a isClosed() check here ?  (now against the AtomicBoolean)

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporters.java [hunk 2]
```diff
@@ -225,7 +233,7 @@ private void doExport(final ExportBulk bulk, final Collection<MonitoringDoc> doc
         } catch (ExportException e) {
             exceptionRef.set(e);
         } finally {
-            bulk.close(lifecycleState() == Lifecycle.State.STARTED, ActionListener.wrap(r -> {
```
Comment (jakelandis) [inline x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporters.java]: there are alot of states here, so i am kinda getting lost... but i _think_ this different then before since this can now flush if the Lifecycle.State.STARTED == false. Not sure if this an issue

**Comment by jakelandis** [comment]
Overall, I think this looks good. However, it is pretty hard to separate the refactoring from the issue this addresses and the refactorings are kinda hard to follow if they are semantically equivalent (at least for me). 

I left a couple questions in the review.

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalBulk.java (hunk 1 above)
Comment (pickypg) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalBulk.java]: We do not. I removed code that cheated and called the implementations directly, which makes this whole method guarded by `add`. Also, the whole `add-flush` flow was always synchronous so we never really made use of the inner-loop escape.

File: x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporters.java (hunk 2 above)
Comment (pickypg) [reply x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporters.java]: You're correct. This is a subtle change in behavior, but it was always an unnecessary one. The clients used by the exporter implementations will be closed when the process shuts down even if we're in the middle of a flush / close, and the `exporter` don't exist until the lifecycle is started, which means we couldn't have even gotten here anyway.

**Comment by pickypg** [comment]
> it is pretty hard to separate the refactoring from the issue this addresses and the refactorings are kinda hard to follow if they are semantically equivalent (at least for me).

Sorry about that. I had also been confused when I revisited the old `close` / `doClose` code to the point that I wondered if the bug was in that code. It was after making that change and further debugging (after I dropped that code and updated tests to handle it) that I discovered the core change ( https://github.com/elastic/elasticsearch/pull/40920/files#diff-a8c42761f9a33243a3256b61442318a6R185 ) that fixes the bug.

I could split these out if you'd prefer, but it sounds like you've got a handle on it now.

**Review by jakelandis — APPROVED** [review]
LGTM thanks for the additional explanations.

**Comment by pickypg** [comment]
7.x: https://github.com/elastic/elasticsearch/commit/64a05e522f1911aa8cb38a2f6c4fb018ff0850fa (https://github.com/elastic/elasticsearch/pull/41043)
7.0: https://github.com/elastic/elasticsearch/commit/42ed969a022f4d58b946ee6105704e8d838c580f (https://github.com/elastic/elasticsearch/pull/41044)
6.7: https://github.com/elastic/elasticsearch/commit/1226a1f2fd05a92e19cd3a0a59e1370e06d47438 (https://github.com/elastic/elasticsearch/pull/41045)
